File: sandbox/app/persistence/repositories/auth_credential_repository.py
"""MySQL repository for browser auth credentials (register/login)."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from sandbox.app.persistence.repositories._base import require_db
from sandbox.security.ownership import BOOTSTRAP_ORG_ID


def _utcnow() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def _row(row: dict[str, Any] | None) -> dict[str, Any] | None:
    if not row:
        return None
    return {
        "id": row["external_user_id"],
        "username": row["username"],
        "password_hash": row["password_hash"],
        "email": row.get("email"),
        "display_name": row.get("display_name"),
        "role": row.get("role") or "user",
        "organization_id": row.get("external_org_id") or BOOTSTRAP_ORG_ID,
        "is_active": bool(row.get("is_active", True)),
        "created_at": row.get("created_at"),
        "updated_at": row.get("updated_at"),
        "last_login_at": row.get("last_login_at"),
    }


class AuthCredentialRepository:
    """CRUD for auth_credentials (Agent-owned schema, Sandbox reads/writes rows)."""

    def __init__(self, db: Any) -> None:
        self.db = require_db(db, "AuthCredentialRepository")

    def create(
        self,
        *,
        username: str,
        password_hash: str,
        external_user_id: str,
        external_org_id: str | None = None,
        email: str | None = None,
        display_name: str | None = None,
        role: str = "user",
    ) -> dict[str, Any]:
        now = _utcnow()
        org_id = (external_org_id or BOOTSTRAP_ORG_ID).strip() or BOOTSTRAP_ORG_ID
        with self.db.connection() as conn:
            conn.execute(
                """
                INSERT INTO auth_credentials (
                  username, password_hash, external_user_id, external_org_id,
                  display_name, email, role, is_active,
                  created_at, updated_at, last_login_at
                ) VALUES (
                  %s, %s, %s, %s, %s, %s, %s, 1, %s, %s, NULL
                )
                """,
                (
                    username,
                    password_hash,
                    external_user_id,
                    org_id,
                    display_name or username,
                    email,
                    role or "user",
                    now,
                    now,
                ),
            )
            conn.commit()
        entry = self.get_by_username(username)
        if not entry:
            raise RuntimeError("auth credential insert did not persist")
        return entry

    def get_by_username(self, username: str) -> dict[str, Any] | None:
        with self.db.connection() as conn:
            conn.execute(
                "SELECT * FROM auth_credentials WHERE username = %s LIMIT 1",
                (username,),
            )
            row = conn.fetchone()
        return _row(row)

    def get_by_external_user_id(self, external_user_id: str) -> dict[str, Any] | None:
        with self.db.connection() as conn:
            conn.execute(
                "SELECT * FROM auth_credentials WHERE external_user_id = %s LIMIT 1",
                (external_user_id,),
            )
            row = conn.fetchone()
        return _row(row)

    def touch_login(self, external_user_id: str) -> None:
        now = _utcnow()
        with self.db.connection() as conn:
            conn.execute(
                """
                UPDATE auth_credentials
                SET last_login_at = %s, updated_at = %s
                WHERE external_user_id = %s
                """,
                (now, now, external_user_id),
            )
            conn.commit()

File: agent/src/infrastructure/mysql/repositories/auth-credential-repository.ts
import { formatDateTime, toMysqlDateTime } from '../row-mappers.js';

type Loose = any;

function mapCredential(row: Record<string, unknown> | undefined) {
  if (!row) return null;
  return {
    id: String(row.external_user_id),
    username: String(row.username),
    passwordHash: String(row.password_hash),
    email: row.email == null ? null : String(row.email),
    displayName: row.display_name == null ? null : String(row.display_name),
    role: String(row.role || 'user'),
    organizationId: String(row.external_org_id || 'org_bootstrap'),
    isActive: Boolean(row.is_active),
    createdAt: formatDateTime(row.created_at),
    lastLoginAt: formatDateTime(row.last_login_at),
  };
}

/** Password credentials live in Agent's MySQL schema, beside identity mappings. */
export class AuthCredentialRepository {
  db: Loose;
  now: () => Date;

  constructor(db: Loose, { now = () => new Date() } = {}) {
    if (!db) throw new Error('AuthCredentialRepository requires a knex executor');
    this.db = db;
    this.now = now;
  }

  async create(input: {
    username: string;
    passwordHash: string;
    externalUserId: string;
    externalOrgId: string;
    email: string | null;
    displayName: string | null;
    role: string;
  }) {
    const now = toMysqlDateTime(this.now());
    await this.db('tbl_agsvc_auth_credentials').insert({
      username: input.username,
      password_hash: input.passwordHash,
      external_user_id: input.externalUserId,
      external_org_id: input.externalOrgId,
      email: input.email,
      display_name: input.displayName || input.username,
      role: input.role,
      is_active: true,
      created_at: now,
      updated_at: now,
      last_login_at: null,
    });
    return this.getByUsername(input.username);
  }

  async getByUsername(username: string) {
    return mapCredential(
      await this.db('tbl_agsvc_auth_credentials').where({ username }).first(),
    );
  }

  async getByExternalUserId(externalUserId: string) {
    return mapCredential(
      await this.db('tbl_agsvc_auth_credentials')
        .where({ external_user_id: externalUserId })
        .first(),
    );
  }

  async setRole(externalUserId: string, role: string) {
    await this.db('tbl_agsvc_auth_credentials')
      .where({ external_user_id: externalUserId })
      .update({ role, updated_at: toMysqlDateTime(this.now()) });
  }

  async touchLogin(externalUserId: string) {
    const now = toMysqlDateTime(this.now());
    await this.db('tbl_agsvc_auth_credentials')
      .where({ external_user_id: externalUserId })
      .update({ last_login_at: now, updated_at: now });
  }

  /**
   * 修改本人的显示名称 / 邮箱。`auth_credentials` 与 `users` 各存一份（后者是
   * 运行账本与通知收件人的来源），同一事务里一起改，不留下一半的状态。
   * `patch` 里没出现的键保持不变。
   */
  async updateProfile(
    externalUserId: string,
    userExternalSubject: string,
    patch: { displayName?: string; email?: string | null },
  ) {
    const now = toMysqlDateTime(this.now());
    const fields: Record<string, unknown> = {};
    if (patch.displayName !== undefined) fields.display_name = patch.displayName;
    if (patch.email !== undefined) fields.email = patch.email;
    if (Object.keys(fields).length) {
      await this.db.transaction(async (trx: Loose) => {
        await trx('tbl_agsvc_auth_credentials')
          .where({ external_user_id: externalUserId })
          .update({ ...fields, updated_at: now });
        await trx('tbl_agsvc_users')
          .where({ external_subject: userExternalSubject })
          .update({ ...fields, updated_at: now });
      });
    }
    return this.getByExternalUserId(externalUserId);
  }
}
